d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -122,6 +161,37 @@ String getXdsServerAuthority()
     return _managedChannel.authority();
   }
 
+  /**
+   * The client may be in backoff if there are RPC stream failures, and if it's waiting to establish the stream again.
+   * NOTE: Must be called from the executor.
+   * @return {@code true} if the client is in backoff
+   */
+  private boolean isInBackoff() {
+    return _adsStream == null && _retryRpcStreamFuture != null && !_retryRpcStreamFuture.isDone();
+  }
+
+  /**
+   * Handles ready callbacks from the RPC stream. Must be called from the executor.
+   */
+  private void readyHandler() {
+    _log.debug("Received ready callback from the ADS stream");
+    if (_adsStream == null || isInBackoff()) {
+      _log.warn("Unexpected state, ready called on null or backed off ADS stream!");
+      return;
+    }
+    // confirm ready state to neglect spurious callbacks; we'll get another callback whenever it is ready again.
+    if (_adsStream.isReady()) {
+      // if the ready timeout future is non-null, a reconnect notification hasn't been sent yet.
+      if (_readyTimeoutFuture != null) {
+        // timeout task will be cancelled only if it hasn't already executed.
+        boolean cancelledTimeout = _readyTimeoutFuture.cancel(false);
+        _log.info("ADS stream ready, cancelled timeout task: {}", cancelledTimeout);
+        _readyTimeoutFuture = null; // set it to null to avoid repeat notifications to subscribers.
+        notifyStreamReconnect();
```

Comment:
Technically, the timeout task should never be executing right now since the executor only has 1 thread.
Just in case there are more threads in the future, why not just interrupt the timeout task if it happens to be executing? Otherwise, even setting the future to null, the task will still complete executing, which could compete with line 190. The subscribers could get notified wrongly with reconnect first, then errors, leaving the _isAvailable in XdsToD2PropertiesAdaptor being false, while the stream is actually reconnected.

Reply:
Most of the code in this class is thread-unsafe and designed to work with a single-threaded executor only, and yes, it's correct that the timeout task shouldn't be executing in parallel. I did consider cancelling using interrupt, but was worried that that might cause unexpected behavior due to the thrown interrupted exception that we're not handling. In contrast, letting the timeout task execute seemed safer. Let me know what you think.

Reply:
yeah, since we just have a single thread currently, we can save the multi-thread related concern for later.